pandas/_libs/groupby.pyx

```diff
@@ -590,10 +604,19 @@ def group_sum(
                     # With dt64/td64 values, values have been cast to float64
                     #  instead if int64 for group_sum, but the logic
                     #  is otherwise the same as in _treat_as_na
-                    if val == val and not (
-                        sum_t is float64_t
+                    if uses_mask:
+                        isna_entry = mask[i, j]
+                    elif (sum_t is float32_t or sum_t is float64_t
+                        or sum_t is complex64_t or sum_t is complex64_t):
+                        # avoid warnings because of equality comparison
+                        isna_entry = not val == val
+                    else:
+                        isna_entry = False
+
+                    if not isna_entry and not (
+                        sum_t is int64_t
                         and is_datetimelike
-                        and val == <float64_t>NPY_NAT
+                        and val == NPY_NAT
```

Comment:
i think this should be part of isna_entry

Reply:
good point, makes way more sense. Thanks